File: Cycle_in_graph.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
using namespace std;

class Graph {
public:
    unordered_map<char, vector<char>> adjList;

    // Add edge to the graph
    void addEdge(char u, char v) {
        adjList[u].push_back(v);
    }

    // Helper function for DFS traversal
    bool dfs(char node, unordered_map<char, bool>& visited, unordered_map<char, bool>& recStack) {
        // Mark the current node as visited and add to the recursion stack
        visited[node] = true;
        recStack[node] = true;

        // Recur for all the adjacent vertices of this node
        for (char neighbor : adjList[node]) {
            // If the neighbor is not visited, recurse on it
            if (!visited[neighbor] && dfs(neighbor, visited, recStack))
                return true;
            // If the neighbor is in the recursion stack, it means there's a cycle
            else if (recStack[neighbor]) {
                return true;
            }
        }

        // Remove the node from recursion stack before returning
        recStack[node] = false;
        return false;
    }

    // Function to detect cycle in the graph
    bool detectCycle() {
        unordered_map<char, bool> visited;
        unordered_map<char, bool> recStack;

        // Check for cycle in every vertex
        for (auto& pair : adjList) {
            if (!visited[pair.first]) {
                if (dfs(pair.first, visited, recStack))
                    return true;
            }
        }
        return false;
    }
};

int main() {
    Graph g;

    // Create a graph with edges
    g.addEdge('A', 'B');
    g.addEdge('B', 'C');
    g.addEdge('C', 'A'); // This creates a cycle: A -> B -> C -> A

    if (g.detectCycle())
        cout << "Cycle detected!" << endl;
    else
        cout << "No cycle detected." << endl;

    return 0;
}

File: cycle_in_graph.py
class Graph:
    def __init__(self):
        self.adj_list = {}

    # Add edge to the graph
    def add_edge(self, u, v):
        if u not in self.adj_list:
            self.adj_list[u] = []
        self.adj_list[u].append(v)

    # Helper function for DFS traversal
    def dfs(self, node, visited, rec_stack):
        # Mark the current node as visited and add to the recursion stack
        visited[node] = True
        rec_stack[node] = True

        # Recur for all the vertices adjacent to this node
        for neighbor in self.adj_list.get(node, []):
            # If the neighbor is not visited, recurse on it
            if not visited.get(neighbor, False) and self.dfs(neighbor, visited, rec_stack):
                return True
            # If the neighbor is in the recursion stack, a cycle is detected
            elif rec_stack.get(neighbor, False):
                return True

        # Remove the vertex from recursion stack before returning
        rec_stack[node] = False
        return False

    # Function to detect cycle in the graph
    def detect_cycle(self):
        visited = {}
        rec_stack = {}

        # Check for cycle in every vertex
        for node in self.adj_list:
            if not visited.get(node, False):
                if self.dfs(node, visited, rec_stack):
                    return True
        return False

# Example usage
g = Graph()

g.add_edge('A', 'B')
g.add_edge('B', 'C')
g.add_edge('C', 'A')  # This creates a cycle: A -> B -> C -> A
# Example for disconnected graphs
g.add_edge('D', 'E')
g.add_edge('E', 'F')  # No cycle here

if g.detect_cycle():
    print("Cycle detected!")
else:
    print("No cycle detected.")
